File: src/app/model/Metric.ts
import { IMetricService } from '../services/IMetricService';
import { UnitConversion } from '../util/unit-conversion';

export enum MetricType {
    ALTITUDE = "ALTITUDE",
    CADENCE = "CADENCE",
    DISTANCE = "DISTANCE",
    GRADE = "GRADE",
    HEART_RATE = "HEART_RATE",
    LATITUDE = "LATITUDE",
    LONGITUDE = "LONGITUDE",
    POWER_BALENCE = "POWER_BALENCE",
    POWER = "POWER",
    SPEED = "SPEED",
    TEMPERATURE = "TEMPERATURE",
    VERTICAL_SPEED = "VERTICAL_SPEED",
    WHEEL_ROTATIONS = "WHEEL_ROTATIONS",
    BEARING = "BEARING"
}


export abstract class Metric {
    protected type: MetricType;
    protected name: string;
    protected siUnit: string;
    protected preferredUnit: string;
    protected preferredPrecision: number;

    protected values: number[];
    protected timestamps: Date[];

    protected metricService: IMetricService;

    constructor(type: MetricType, name: string, siUnit: string, metricService: IMetricService, preferredUnit: string = siUnit, preferredPrecision: number = 0) {
        this.type = type;
        this.metricService = metricService;
        this.name = name;
        this.siUnit = siUnit;
        this.preferredUnit = preferredUnit;
        this.preferredPrecision = preferredPrecision;
        this.values = [];
        this.timestamps = [];
    }

    /**
     * Start logging the metric. This method should be implemented by subclasses to start the logging process.
     */
    abstract startLogging(): void;

    /**
     * Stop logging the metric. This method should be implemented by subclasses to stop the logging process.
     */
    abstract stopLogging(): void;

    addValue(value: number, timestamp: Date) {
        this.values.push(value);
        this.timestamps.push(timestamp);
    }
    getValues(): number[] {
        return this.values;
    }
    getTimestamps(): Date[] {
        return this.timestamps;
    }
    getName(): string {
        return this.name;
    }
    getPreferredUnit(): string {
        return this.preferredUnit;
    }
    getLastValue(): number | null {
        if (this.values.length > 0) {
            return this.values[this.values.length - 1];
        }
        return null;
    }

    get3sAverage(): number | null {
        const values = this.getValuesUnil(3);
        if (values.length === 0) {
            return null;
        }
        const sum = values.reduce((acc, value) => acc + value, 0);
        return sum / values.length;
    }

    get30sAverage(): number | null {
        const values = this.getValuesUnil(30);
        if (values.length === 0) {
            return null;
        }
        const sum = values.reduce((acc, value) => acc + value, 0);
        return sum / values.length;
    }

    getAverage(): number | null {
        if (this.values.length === 0) {
            return null;
        }
        const sum = this.values.reduce((acc, value) => acc + value, 0);
        return sum / this.values.length;
    }

    getMax(): number | null {
        if (this.values.length === 0) {
            return null;
        }
        return Math.max(...this.values);
    }


    displayLastValue(includeUnit: boolean = true): string {
        const lastValue = this.getLastValue();
        return this.displayValue(lastValue, includeUnit);
    }

    display3sAverage(includeUnit: boolean = true): string {
        const average = this.get3sAverage();
        return this.displayValue(average, includeUnit);
    }

    display30sAverage(includeUnit: boolean = true): string {
        const average = this.get30sAverage();
        return this.displayValue(average, includeUnit);
    }

    displayAverage(includeUnit: boolean = true): string {
        const average = this.getAverage();
        return this.displayValue(average, includeUnit);
    }

    displayMax(includeUnit: boolean = true): string {
        const max = this.getMax();
        return this.displayValue(max, includeUnit);
    }

    getLastTimestamp(): Date | null {
        if (this.timestamps.length > 0) {
            return this.timestamps[this.timestamps.length - 1];
        }
        return null;
    }
    getType(): MetricType {
        return this.type;
    }

    getValuesUnil(secondsAgo: number): number[] {
        const now = new Date();
        const threshold = new Date(now.getTime() - secondsAgo * 1000);
        return this.values.filter((_, index) => this.timestamps[index] >= threshold);
    }

    private displayValue(value: number | null, includeUnit: boolean = true): string {
        if (value === null) {
            return '---';
        }
        let valueConverted = value;
        if (this.siUnit !== this.preferredUnit) {
            valueConverted = UnitConversion.convert(value, this.siUnit, this.preferredUnit);
        }
        if (includeUnit) {
            return `${valueConverted.toFixed(this.preferredPrecision)} ${this.preferredUnit}`;
        }
        return `${valueConverted.toFixed(this.preferredPrecision)}`;
    }
}

File: android/app/src/main/java/com/openbikecomputer/model/Metric.java
package com.openbikecomputer.model;

import com.openbikecomputer.service.IMetricService;
import com.openbikecomputer.util.UnitConversion;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

/**
 * Base class for all metrics. Stores a time series of values and exposes
 * averaging / formatting helpers. Ported from {@code Metric.ts}.
 *
 * <p>Timestamps are epoch milliseconds (the Angular app used {@code Date}).</p>
 */
public abstract class Metric {
    protected final MetricType type;
    protected final String name;
    protected final String siUnit;
    protected final String preferredUnit;
    protected final int preferredPrecision;

    protected final List<Double> values = new ArrayList<>();
    protected final List<Long> timestamps = new ArrayList<>();
    /** Starting index of each lap. */
    protected final List<Integer> laps = new ArrayList<>();

    protected final IMetricService metricService;

    protected Metric(MetricType type, String name, String siUnit, IMetricService metricService) {
        this(type, name, siUnit, metricService, siUnit, 0);
    }

    protected Metric(MetricType type, String name, String siUnit, IMetricService metricService,
                     String preferredUnit) {
        this(type, name, siUnit, metricService, preferredUnit, 0);
    }

    protected Metric(MetricType type, String name, String siUnit, IMetricService metricService,
                     String preferredUnit, int preferredPrecision) {
        this.type = type;
        this.name = name;
        this.siUnit = siUnit;
        this.metricService = metricService;
        this.preferredUnit = preferredUnit;
        this.preferredPrecision = preferredPrecision;
        this.laps.add(0);
    }

    /** Start logging the metric (subscribe to sensors / timers). */
    public abstract void startLogging();

    /** Stop logging the metric. */
    public abstract void stopLogging();

    public void newLap() {
        laps.add(values.size());
    }

    public void addValue(double value, long timestamp) {
        values.add(value);
        timestamps.add(timestamp);
    }

    public List<Double> getValues() {
        return values;
    }

    public List<Long> getTimestamps() {
        return timestamps;
    }

    public String getName() {
        return name;
    }

    public String getPreferredUnit() {
        return preferredUnit;
    }

    public MetricType getType() {
        return type;
    }

    public Double getLastValue() {
        if (!values.isEmpty()) {
            return values.get(values.size() - 1);
        }
        return null;
    }

    public Double get3sAverage() {
        return average(getValuesUntil(3));
    }

    public Double get30sAverage() {
        return average(getValuesUntil(30));
    }

    public Double getAverage() {
        return average(values);
    }

    public Double getMax() {
        if (values.isEmpty()) {
            return null;
        }
        double max = Double.NEGATIVE_INFINITY;
        for (double v : values) {
            if (v > max) {
                max = v;
            }
        }
        return max;
    }

    public Double getAverageForLap(Integer lapIndex) {
        if (lapIndex == null) {
            lapIndex = laps.size() - 1;
        }
        if (lapIndex < 0 || lapIndex >= laps.size()) {
            throw new IllegalArgumentException("Invalid lap index: " + lapIndex);
        }
        int start = laps.get(lapIndex);
        int end = lapIndex + 1 < laps.size() ? laps.get(lapIndex + 1) : values.size();
        if (start >= end) {
            return null;
        }
        return average(values.subList(start, end));
    }

    public String displayLastValue(boolean includeUnit) {
        return displayValue(getLastValue(), includeUnit);
    }

    public String displayLastValue() {
        return displayLastValue(true);
    }

    public String display3sAverage(boolean includeUnit) {
        return displayValue(get3sAverage(), includeUnit);
    }

    public String display3sAverage() {
        return display3sAverage(true);
    }

    public String display30sAverage(boolean includeUnit) {
        return displayValue(get30sAverage(), includeUnit);
    }

    public String display30sAverage() {
        return display30sAverage(true);
    }

    public String displayAverage(boolean includeUnit) {
        return displayValue(getAverage(), includeUnit);
    }

    public String displayMax(boolean includeUnit) {
        return displayValue(getMax(), includeUnit);
    }

    public String displayAverageForLap(Integer lapIndex, boolean includeUnit) {
        return displayValue(getAverageForLap(lapIndex), includeUnit);
    }

    public String displayAverageForLap(Integer lapIndex) {
        return displayAverageForLap(lapIndex, true);
    }

    public String displayAverageForLap() {
        return displayAverageForLap(null, true);
    }

    public Long getLastTimestamp() {
        if (!timestamps.isEmpty()) {
            return timestamps.get(timestamps.size() - 1);
        }
        return null;
    }

    /** Values logged within the last {@code secondsAgo} seconds. */
    protected List<Double> getValuesUntil(double secondsAgo) {
        long threshold = System.currentTimeMillis() - (long) (secondsAgo * 1000);
        List<Double> result = new ArrayList<>();
        for (int i = 0; i < values.size(); i++) {
            if (timestamps.get(i) >= threshold) {
                result.add(values.get(i));
            }
        }
        return result;
    }

    private static Double average(List<Double> input) {
        if (input == null || input.isEmpty()) {
            return null;
        }
        double sum = 0;
        for (double v : input) {
            sum += v;
        }
        return sum / input.size();
    }

    protected String displayValue(Double value, boolean includeUnit) {
        if (value == null) {
            return "00.00";
        }
        double valueConverted = value;
        if (!siUnit.equals(preferredUnit)) {
            valueConverted = UnitConversion.convert(value, siUnit, preferredUnit);
        }
        String formatted = String.format(Locale.US, "%." + preferredPrecision + "f", valueConverted);
        if (includeUnit) {
            return formatted + " " + preferredUnit;
        }
        return formatted;
    }
}
